### PR TITLE
"Become Avatar" usefulUtility and JAT's script

### PR DESCRIPTION
**Pull request by zfox23**
JAT's script has been "QA'd". The "Become Avatar" script hasn't - a quick pass on that would be nice! Use any FST model, then attach the script.

### CODE REVIEW

**Comment by AndySeattle** [comment]
I have the click me script staged on some avatars now, that looks good. 

Question though, could we set this up to have the avatar list be part of the config.json instead of hard coded in the script? or would that be to heavy of a lift? Because then we are a stones throw away from my dream for this sort of app showing my favorite avatars as a clickable avatar entity that i only see. (i know that second is a larger ask, but i feel like doing the groundwork now would not be a crazy ask? or would it?)

**Comment by zfox23** [comment]
@AndySeattle:

>could we set this up to have the avatar list be part of the config.json instead of hard coded in the script?

Could you please explain what this means? I don't understand. It feels like this might be a separate feature request, but I don't yet understand the request.

**Comment by AndySeattle** [comment]
> @AndySeattle:
> 
> > could we set this up to have the avatar list be part of the config.json instead of hard coded in the script?
> 
> Could you please explain what this means? I don't understand. It feels like this might be a separate feature request, but I don't yet understand the request.

Nevermind you are already doing what i want it to do, the other is indeed a feature request, well actually a separate thing altogether. (A way to visualize your avatars in a 3d space.) I just reviewed the files for the avatar selector and the blank config file, did not fully read the readme for the JAT script :) what you are doing is what i wanted you to do :) So i am gold. about to rig this with some of my avatars and give it a go. 

**my BIG question** however looking at the code, could we reference the avatar locations in the config file instead of hard coding them? That would make this much more malleable. (as we are already pulling the avatar FST files from a reference in the config.json, could we also not pull the locations? I feel like having them hardcoded in the script, but the avatars not hardcoded in the script is disjoint of the data pertaining to them. )

**Comment by zfox23** [comment]
Thank you for QAing this! And I appreciate your comments. What you're asking for (which is a great suggestion) belongs in a separate (small) project rather than in James Arnold Taylor's script specifically.

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+//  Copyright 2019 High Fidelity, Inc.
+//
+//  See accompanying README.md for usage instructions.
+// 
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+
+
+(function () {
+    var BASE_RELATIVE_POSITION = {"x":20.9303035736084,"y":-11.449395179748535,"z":-50.0930061340332};
```
Comment (robin-k-wilson) [inline events/multiCon/jamesArnoldTaylor/JATScript.js]: Could we put this in config? If there's no BASE_RELATIVE_POSITION defined, we could simply use MyAvatar.postion. I feel like this hard coded value is specific to one domain and will be changed in the future. MyAvatar.position is a nice default.

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    ];
+
+    var BUTTON_DIMENSIONS = {"x": 0.2, "y": 0.2, "z": 0.2};
+    var BUTTON_ROTATION = Quat.fromPitchYawRollDegrees(180, 140, 180);
+    var BUTTON_ROTATION_PREV = Quat.fromPitchYawRollDegrees(180, -140, 0);
+
+    var CONTROL_BUTTON_PROPERTIES = [
+        {
+            "name": "play",
+            "url": "https://hifi-content.s3.amazonaws.com/alan/dev/playback_play-button.fbx",
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: that's a fantastic idea!

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        "dpi": 1920 / (3 * METERS_TO_INCHES),
+        "alpha": 0.75
+    };
+
+    var rezzedAvatarOverlays = [];
+    function rezAvatarOverlays() {
+        for (var i = 0; i < AVATAR_OVERLAY_PROPERTIES.length; i++) {
+            var currentProps = AVATAR_OVERLAY_PROPERTIES[i];
+            currentProps.position = Vec3.sum(currentProps.position, BASE_RELATIVE_POSITION);
+            console.log("Rezzing overlay with name " +
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: logs aren't always necessary to remove. in this case, this is a script that is only running on one computer (probably) that is used to assist a performer. if something goes wrong during an event, we want as much help as we can get to try to diagnose the problem - and thus logs can be a big help!

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                AVATAR_OVERLAY_PROPERTIES[i].name + " at " + JSON.stringify(currentProps.position));
+            rezzedAvatarOverlays.push(Overlays.addOverlay("model", currentProps));
+        }
+    }
+
+    var rezzedControlButtonOverlays = [];
+    function rezVideoControlButtons() {
+        for (var i = 0; i < CONTROL_BUTTON_PROPERTIES.length; i++) {
+            var currentProps = CONTROL_BUTTON_PROPERTIES[i];
+            console.log("Rezzing overlay with name " +
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                CONTROL_BUTTON_PROPERTIES[i].name + " at " + JSON.stringify(currentProps.position));
+            rezzedControlButtonOverlays.push(Overlays.addOverlay("model", currentProps));
+        }
+    }
+
+    var textOverlay = false;
+    function rezTextOverlay() {
+        console.log("Rezzing overlay with name " +
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            TEXT_OVERLAY_PROPERTIES.name + " at " + JSON.stringify(TEXT_OVERLAY_PROPERTIES.position));
+        textOverlay = Overlays.addOverlay("text3d", TEXT_OVERLAY_PROPERTIES);
+    }
+
+    var webOverlay = false;
+    function rezWebOverlay() {
+        console.log("Rezzing overlay with name " +
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        WEB_OVERLAY_PROPERTIES.name + " at " + JSON.stringify(WEB_OVERLAY_PROPERTIES.position));
+        webOverlay = Overlays.addOverlay("web3d", WEB_OVERLAY_PROPERTIES);
+    }
+
+    function startup() {
+        console.log("Welcome to rezJATAvatars.js!");
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: 😂 see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+    
+        rezAvatarOverlays();
+        rezVideoControlButtons();
+        rezTextOverlay();
+        rezWebOverlay();
+    }
+
+    function onScriptEnding() {
+        for (var i = 0; i < rezzedAvatarOverlays.length; i++) {
+            console.log("Deleting local entity with ID " + rezzedAvatarOverlays[i]);
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+            Overlays.deleteOverlay(rezzedAvatarOverlays[i]);
+        }
+        for (i = 0; i < rezzedAvatarOverlays.length; i++) {
+            console.log("Deleting local entity with ID " + rezzedControlButtonOverlays[i]);
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+        }
+
+        if (textOverlay) {
+            Overlays.deleteOverlay(textOverlay);
+        }
+
+        if (webOverlay) {
+            Overlays.deleteOverlay(webOverlay);
+        }
+
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: done!

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+            default:
+                console.log("Unhandled button overlay pressed");
+        }
+    }
+
+    var STATUS_MESSAGE_CHANNEL = config.statusMessageChannel;
+    function onMessageReceived(channel, message) {
+        if (channel === STATUS_MESSAGE_CHANNEL && textOverlay){
+            console.log("message", message);
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: see comment above

File: events/multiCon/jamesArnoldTaylor/JATScript.js [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        
+        if (rezzedAvatarOverlays.indexOf(id) > -1) {
+            var modelURL = Overlays.getProperties(id, ["url"]).url;
+            MyAvatar.useFullAvatarURL(modelURL);
+        } else if (rezzedControlButtonOverlays.indexOf(id) > -1) {
+            handleControlButtonPress(id);
+        }
+    }
+
+    Overlays.mousePressOnOverlay.connect(onMousePressOnOverlay);
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: good call :)

File: events/multiCon/jamesArnoldTaylor/JATScript.js (hunk 1 above)
Comment (zfox23) [reply events/multiCon/jamesArnoldTaylor/JATScript.js]: this hard coded value _is_ specific to one domain and _will_ be changed in the future - this is a script specific to an event, and specific to one person. i don't think it's necessary _for this script_ to put something like this in the config file.

**Pull request merged**